lib/client.js

```diff
@@ -219,12 +220,37 @@ MqttClient.prototype._setupStream = function () {
   // Echo connection errors
   parser.on('error', this.emit.bind(this, 'error'));
 
-  this.outgoingStore
-    .createStream()
-    .on('data', function (packet) {
-      that._sendPacket(packet);
+  outStore = this.outgoingStore
+  .createStream();
+  // Control of stored messages
+  outStore.once('readable', function () {
+      function storeDeliver () {
+        var packet = outStore.read(1);
+        if (!packet) {
+          return;
+        }
+        // Avoid unnecesary stream read operations
+        if (!that.disconnecting && !that.reconnectTimer) {
+          outStore.read(0);
+          // Ensure that the next message will only be read after callback is issued
+          that.outgoing[packet.messageId] = function () {
+            storeDeliver();
+          };
+          that._sendPacket(packet);
+        }
```

Comment:
this does not close the stream if the client gets disconnected, it might leak memory.